sycl/doc/extensions/supported/sycl_ext_oneapi_bfloat16.asciidoc

```diff
@@ -0,0 +1,311 @@
+`sycl::half` and 7 bits to the significand versus 10 bits used by `sycl::half`.
+Thus, `bfloat16` has the same dynamic range as a 32-bit `float` but with
+reduced precision. This type is useful when memory required to store the values
+must be reduced, and when the calculations require high dynamic range but can
+tolerate lower-precision. Some implementations may still perform operations
+ on this type using 32-bit math. For example, they may convert the `bfloat16`
+ value to `float`, and then perform the operation on the 32-bit `float`.
+
+[NOTE]
+The bfloat16 type is supported on all devices. DPC++ currently supports this
+type natively on Intel Xe HP GPUs and Nvidia GPUs with
+Compute Capability >= SM80. On other devices it is emulated in software.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros". Therefore, an implementation
+supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_BFLOAT16` to one of the values defined in the table
+below. Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro’s
+value to determine which of the extension’s APIs the implementation supports.
+
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version. Base features are supported.
+|===
+
+
+=== New `bfloat16` class
+
+The `bfloat16` type represents a 16-bit floating point value.
+Conversions from `float` to `bfloat16` are done with round to
+nearest even (RTE) rounding mode.
+
+[source]
+----
+namespace sycl {
+namespace ext {
+namespace oneapi {
+
+class bfloat16 {
+
+public:
+  bfloat16() = default;
+  bfloat16(const bfloat16 &) = default;
+  ~bfloat16() = default;
+
+  // Convert from float to bfloat16
+  bfloat16(const float &a);
+  bfloat16 &operator=(const float &a);
+
+  // Convert bfloat16 to float
+  operator float() const;
+  
+  // Convert from sycl::half to bfloat16
+  bfloat16(const sycl::half &a);
+  bfloat16 &operator=(const sycl::half &a);
+
+  // Convert bfloat16 to sycl::half
+  operator sycl::half() const;
+
+  // Convert bfloat16 to bool type
+  explicit operator bool();
+
+  friend bfloat16 operator-(bfloat16 &bf) { /* ... */ }
+
+  // OP is: prefix ++, --
+  friend bfloat16 &operatorOP(bfloat16 &bf) { /* ... */ }
+
+  // OP is: postfix ++, --
+  friend bfloat16 operatorOP(bfloat16 &bf, int) { /* ... */ }
+
+  // OP is: +=, -=, *=, /=
+  friend bfloat16 &operatorOP(bfloat16 &lhs, const bfloat16 &rhs) { /* ... */ }
+
+  // OP is +, -, *, /
+  friend bfloat16 operatorOP(const bfloat16 &lhs, const bfloat16 &rhs)
+  { /* ... */ }
+  template <typename T>
+  friend bfloat16 operatorOP(const bfloat16 &lhs, const T &rhs) { /* ... */ }
+  template <typename T>
+  friend bfloat16 operatorOP(const T &lhs, const bfloat16 &rhs) { /* ... */ }
+
+  // OP is ==,!=, <, >, <=, >=
```

Comment:
I am curious. Is it possible to replace these three with a single template function? 
template <typename T1, typename T2> 
friend bool operatorOP(const T1 &lhs, const T2 &rhs) { .. } ?
Looks a lot cleaner. But, i might be missing a finer detail here.

Reply:
The three definitions have at least one input operand explicitly declared as type "bfloat16". This ensures that these functions take effect only for bfloat16 operations.